conditional: add conditional processing documentation (#1598)
fluent/fluent-bit-docs: pipeline/processors/conditional-processing.md

# Conditional processing

> Conditional processing allows you to selectively apply processors to log records based on field values. This feature enables you to create processing pipelines that apply processors only to records that match specific criteria.

Reviewer: does conditional processing ONLY apply to logs? (I see that certain processors apply to metrics/traces as well; not sure if that means that certain types of processors can't be used conditionally since they don't work with logs)

> Conditional processing allows you to selectively apply processors to log records based on field values. This feature enables you to create processing pipelines that apply processors only to records that match specific criteria.

Reviewer: conditional (pun not intended) suggestion for now, but if it turns out that conditional processing also applies to metrics/traces, I'll need to edit this accordingly (and a few other places throughout the doc).
```suggestion
Conditional processing lets you selectively apply [processors](README.md) by determining whether
to process logs based on the value of fields they contain. This feature lets you create processing pipelines
that only process records that meet certain criteria.
```

## Configuration format

Conditional processing is available for processors in the YAML configuration format. To apply a processor conditionally, you add a `condition` block to the processor configuration:

```yaml
- name: processor_name
  # Regular processor configuration...
  condition:
    op: and|or
    rules:
      - field: "$field_name"

>         op: comparison_operator

Reviewer: is it too late to change how this feature is designed so that `condition.op` and `condition.rules.op` have different names? trying to distinguish between the two/explain how they're different is a bit difficult.

(maybe `condition.op` would be `global_operator` and `condition.rules.op` would be `rule_operator`...)

> ## Configuration format
> 
> Conditional processing is available for processors in the YAML configuration format. To apply a processor conditionally, you add a `condition` block to the processor configuration:
> 
> ```yaml
> - name: processor_name
>   # Regular processor configuration...
>   condition:
>     op: and|or
>     rules:
>       - field: "$field_name"
>         op: comparison_operator
>         value: comparison_value
>       # Additional rules...
> ```
> 
> ### Condition operators
> 
> The `op` field in the condition block specifies the logical operator to apply across all rules:
> 
> | Operator | Description |
> | --- | --- |
> | `and` | All rules must evaluate to true for the condition to be true |
> | `or` | At least one rule must evaluate to true for the condition to be true |
> 
> ### Rules
> 
> Each rule consists of:
> 
> - `field`: The field to evaluate (must use [record accessor syntax](/administration/configuring-fluent-bit/classic-mode/record-accessor.md) with `$` prefix)
> - `op`: The comparison operator
> - `value`: The value to compare against
> 
> ### Comparison operators
> 
> The following comparison operators are supported:
> 
> | Operator | Description |
> | --- | --- |
> | `eq` | Equal to |
> | `neq` | Not equal to |
> | `gt` | Greater than |
> | `lt` | Less than |
> | `gte` | Greater than or equal to |
> | `lte` | Less than or equal to |
> | `regex` | Matches regular expression |
> | `not_regex` | Does not match regular expression |
> | `in` | Value is in the specified array |
> | `not_in` | Value is not in the specified array |
> 
> ### Field access
> 
> You can access record fields using [record accessor syntax](/administration/configuring-fluent-bit/classic-mode/record-accessor.md):
> 
> - Simple fields: `$field`
> - Nested fields: `$parent['child']['subchild']`

Reviewer: ```suggestion
## Configuration

You can turn a standard processor into a conditional processor by adding a `condition` block to the
processor's YAML configuration settings.

{% hint style="info" %}
Conditional processing is only available for
[YAML configuration files](../../administration/configuring-fluent-bit/yaml/README.md),
not [classic configuration files](../../administration/configuring-fluent-bit/classic-mode/README.md).
{% endhint %}

These `condition` blocks use the following syntax:

```yaml
pipeline:
  inputs:
  <...>
      processors:
        logs:
          - name: {processor_name}
            <...>
            condition:
	      op: {and|or}
              rules:
                - field: {field_name1}
                  op: {comparison_operator}
                  value: {comparison_value1}
                - field: {field_name2}
                  op: {comparison_operator}
                  value: {comparison_value2}
            <...>
```

Each processor can only have a single `condition` block, but can have multiple rules within that condition.
These rules are stored as items in the `condition.rules` array.

### Condition evaluation

The `condition.op` parameter specifies the condition's evaluation logic. It has two possible values:

- `and`: All rules in the `condition.rules` array must evaluate to `true` for the condition to be met.
- `or`: One or more rules in the `conditions.rules` array must evaluate to `true` for the condition to be met.

### Rule parameters

Each item in the `condition.rules` array must include values for the following parameters:

| Parameter | Description |
| --- | --- |
| `field` | The field within your logs to evaluate. The value of this parameter must use [the correct syntax](#field-access) to access the fields inside logs. |
| `op` | The [comparison operator](#comparison-operators) to evaluate whether the rule is true. This parameter (`condition.rules.op`) is distinct from the `condition.op` parameter and has different possible values. |
| `value` | The value of the specified log field to use in your comparison. | 

#### Field access

The `conditions.rules.field` parameter uses [record accessor syntax](/administration/configuring-fluent-bit/classic-mode/record-accessor.md) to reference fields inside logs.

You can use `$field` syntax to access a top-level field, and `$field['child']['subchild']` to access nested fields.

#### Comparison operators

The `conditions.rules.op` parameter has the following possible values:

- `eq`: equal to
- `neq`: not equal to
- `gt`: greater than
- `lt`: less than
- `gte`: greater than or equal to
- `lte`: less than or equal to
- `regex`: matches a regular expression
- `not_regex`: does not match a regular expression
- `in`: is included in the specified array
- `not_in`: is not included in the specified array
```

## Examples

> ### Simple condition
> 
> Process records only when the HTTP method is POST:

Reviewer: ```suggestion
### Basic condition

This example applies a condition that only processes records with the HTTP method `POST`:
```

      tag: request.log
      processors:
        logs:
          - name: content_modifier
            action: insert
            key: requires_audit
            value: true
            condition:
              op: and
              rules:
                - field: "$request['method']"
                  op: eq
                  value: "POST"
                - field: "$request['path']"
                  op: regex
                  value: "\/sensitive-.*"
```

### OR condition example

Flag records that meet any of multiple criteria:

```yaml
pipeline:
  inputs:
    - name: dummy
      dummy: '{"request": {"method": "GET", "path": "/api/v1/resource", "status_code": 200, "response_time": 150}}'
      tag: request.log
      processors:
        logs:
          - name: content_modifier
            action: insert
            key: requires_performance_check
            value: true
            condition:
              op: or
              rules:
                - field: "$request['response_time']"
                  op: gt
                  value: 100
                - field: "$request['status_code']"
                  op: gte
                  value: 400
```

> ### Using IN operator
> 
> Apply a processor when a value matches one of multiple options:

Reviewer: ```suggestion
### Array of values

This example uses an array for the value of `condition.rules.value`:
```

```yaml
pipeline:
  inputs:
    - name: dummy
      dummy: '{"request": {"method": "GET", "path": "/api/v1/resource"}}'
      tag: request.log
      processors:
        logs:
          - name: content_modifier
            action: insert
            key: high_priority_method
            value: true
            condition:
              op: and
              rules:
                - field: "$request['method']"
                  op: in
                  value: ["POST", "PUT", "DELETE"]
```

> ## Multiple processors with conditions
> 
> You can chain multiple conditional processors to create advanced processing pipelines:

Reviewer: ```suggestion
### Multiple processors with conditions

This example uses multiple processors with conditional processing enabled for each:
```
